crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2673,6 +2722,73 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Performs lexicographic comparison between two slices of types.
+    ///
+    /// This comparison supports only equality and ordering operators: `==`, `!=`, `<`, `<=`, `>`, `>=`.
+    /// For lexicographic comparison, elements from both slices are compared pairwise using
+    /// `infer_binary_type_comparison`. If a conclusive result cannot be determined as a `BoolLiteral`,
+    /// it returns `bool`. If the comparison is undefined between types, it returns `None`.
+    fn infer_lexicographic_type_comparison(
+        &mut self,
+        left: &[Type<'db>],
+        op: ast::CmpOp,
+        right: &[Type<'db>],
+    ) -> Option<Type<'db>> {
+        if !matches!(
+            op,
+            ast::CmpOp::Eq
+                | ast::CmpOp::NotEq
+                | ast::CmpOp::Lt
+                | ast::CmpOp::LtE
+                | ast::CmpOp::Gt
+                | ast::CmpOp::GtE
+        ) {
+            return None;
+        }
+
+        // Compare paired elements from left and right slices
+        for (l_ty, r_ty) in left.iter().zip(right.iter()) {
+            let eq_result_ty = self.infer_binary_type_comparison(*l_ty, ast::CmpOp::Eq, *r_ty);
+
+            match eq_result_ty {
+                // Types are equal, continue to the next pair
+                Some(Type::BooleanLiteral(true)) => continue,
+
+                // Types are not equal, perform the specified comparison and return the result
+                Some(Type::BooleanLiteral(false)) => {
+                    return self.infer_binary_type_comparison(*l_ty, op, *r_ty)
```

Comment:
In this case I think if we end up comparing tuples with non-comparable types in them, as in this example:
```
>>> (1,) < ("foo",)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'int' and 'str'
```

We will give a less informative diagnostic than the runtime does, because we just propagate a `None`, we don't propagate details about which two types were ultimately not orderable. That is, our diagnostic would say something like operator `<` not supported between `tuple[int]` and `tuple[str]`, rather than saying (like runtime does) that the real problem is `int` vs `str`. (It's obvious in a minimal example of course, maybe not in a bigger one.)

The fix for this will to have `infer_binary_type_comparison` return a `Result` rather than `Option`, where the `Err` variant carries with it the two types we could not compare, and then we use this information in the diagnostic.

I think it's out of scope to do that in this PR, but I would like a test for the pass-through-None case here, that checks the diagnostic, and has a TODO for the more informative diagnostic.

Reply:
Yes! If I understood correctly, I’ve added some test cases here:

[MD Test - Comparison Unsupported Part](https://github.com/astral-sh/ruff/pull/13712/files#diff-eff9f162b9d83ddfe43d3f6b28ed022767bc58d12d7ccb0c95d985e827471362R83-R103)